From fac821e139742e954f9a4cf6a7c6f91c33f2e791 Mon Sep 17 00:00:00 2001
From: Jason Wessel <jason.wessel@windriver.com>
Date: Thu, 26 Jan 2017 06:53:28 -0800
Subject: [PATCH 5/6] Allow detached mode to start from systemd

When starting a pflask in the detached mode from systemd the STDIN
will never be a full pty.  In this case any tty operations on STDIN
should be skipped.

Signed-off-by: Jason Wessel <jason.wessel@windriver.com>
---
 src/pty.c | 24 ++++++++++++++++--------
 1 file changed, 16 insertions(+), 8 deletions(-)

diff --git a/src/pty.c b/src/pty.c
index a4ac143..0ca7325 100644
--- a/src/pty.c
+++ b/src/pty.c
@@ -51,6 +51,7 @@
 
 static struct termios stdin_attr;
 static struct winsize stdin_ws;
+static int stdin_is_tty = 1;
 
 static int recv_fd(int sock);
 static void send_fd(int sock, int fd);
@@ -58,11 +59,16 @@ static void send_fd(int sock, int fd);
 void open_master_pty(int *master_fd, char **master_name) {
 	int rc;
 
-	rc = tcgetattr(STDIN_FILENO, &stdin_attr);
-	sys_fail_if(rc < 0, "tcgetattr()");
+	if (isatty(STDIN_FILENO)) {
+		stdin_is_tty = 1;
+		rc = tcgetattr(STDIN_FILENO, &stdin_attr);
+		sys_fail_if(rc < 0, "tcgetattr()");
 
-	rc = ioctl(STDIN_FILENO, TIOCGWINSZ, &stdin_ws);
-	sys_fail_if(rc < 0, "ioctl(TIOCGWINSZ)");
+		rc = ioctl(STDIN_FILENO, TIOCGWINSZ, &stdin_ws);
+		sys_fail_if(rc < 0, "ioctl(TIOCGWINSZ)");
+	} else {
+		stdin_is_tty = 0;
+	}
 
 	*master_fd = posix_openpt(O_RDWR | O_NOCTTY | O_NDELAY);
 	sys_fail_if(*master_fd < 0, "Error opening master pty");
@@ -93,11 +99,13 @@ void open_slave_pty(const char *master_name) {
 	rc = dup2(slave_fd, STDERR_FILENO);
 	sys_fail_if(rc < 0, "dup2(STDERR)");
 
-	rc = tcsetattr(slave_fd, TCSANOW, &stdin_attr);
-	sys_fail_if(rc < 0, "tcsetattr()");
+	if (stdin_is_tty) {
+		rc = tcsetattr(slave_fd, TCSANOW, &stdin_attr);
+		sys_fail_if(rc < 0, "tcsetattr()");
 
-	rc = ioctl(slave_fd, TIOCSWINSZ, &stdin_ws);
-	sys_fail_if(rc < 0, "ioctl(TIOCWINSZ)");
+		rc = ioctl(slave_fd, TIOCSWINSZ, &stdin_ws);
+		sys_fail_if(rc < 0, "ioctl(TIOCWINSZ)");
+	}
 }
 
 void process_pty(int master_fd, char escape_char) {
-- 
2.11.0

